narwhal/config/src/committee.rs

```diff
@@ -0,0 +1,512 @@
+pub struct Authority {
+    /// The id under which we identify this authority across Narwhal
+    #[serde(skip)]
+    id: OnceCell<AuthorityIdentifier>,
+    /// The authority's main PublicKey which is used to verify the content they sign.
+    #[serde(skip)]
+    protocol_key: OnceCell<PublicKey>,
+    /// The authority's main PublicKey expressed as pure bytes
+    #[serde(skip)]
+    protocol_key_bytes: OnceCell<PublicKeyBytes>,
+    /// The voting power of this authority.
+    stake: Stake,
+    /// The network address of the primary.
+    primary_address: Multiaddr,
+    /// Network key of the primary.
+    network_key: NetworkPublicKey,
+}
+
+impl Authority {
+    /// The constructor is not public by design. Everyone who wants to create authorities should do
+    /// it via Committee (more specifically can use CommitteeBuilder). As some internal properties of
+    /// Authority are initialised via the Committee, to ensure that the user will not accidentally use
+    /// stale Authority data, should always derive them via the Commitee.
+    fn new(
+        protocol_key: PublicKey,
+        stake: Stake,
+        primary_address: Multiaddr,
+        network_key: NetworkPublicKey,
+    ) -> Self {
+        let protocol_key_bytes = PublicKeyBytes::from(&protocol_key);
+
+        Self {
+            id: Default::default(),
+            protocol_key: OnceCell::with_value(protocol_key),
+            protocol_key_bytes: OnceCell::with_value(protocol_key_bytes),
+            stake,
+            primary_address,
+            network_key,
+        }
+    }
+
+    pub fn id(&self) -> &AuthorityIdentifier {
+        self.id
+            .get()
+            .expect("Identifier should have been initialised by now")
+    }
+
+    pub fn protocol_key(&self) -> &PublicKey {
+        self.protocol_key
+            .get()
+            .expect("PublicKey should have been initialised by now")
+    }
+
+    pub fn protocol_key_bytes(&self) -> &PublicKeyBytes {
+        self.protocol_key_bytes
+            .get()
+            .expect("ProtocolKeyBytes should have been initialised by now")
+    }
+
+    pub fn stake(&self) -> Stake {
+        self.stake
+    }
+
+    pub fn primary_address(&self) -> Multiaddr {
+        self.primary_address.clone()
+    }
+
+    pub fn network_key(&self) -> NetworkPublicKey {
+        self.network_key.clone()
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug, Eq, PartialEq)]
+pub struct Committee {
+    /// The authorities of epoch.
+    authorities: BTreeMap<PublicKey, Authority>,
+    /// Keeps and index of the Authorities by their respective identifier
+    #[serde(skip)]
+    authorities_by_id: OnceCell<BTreeMap<AuthorityIdentifier, Authority>>,
+    /// The epoch number of this committee
+    epoch: Epoch,
+}
+
+// Every authority gets uniquely identified by the AuthorityIdentifier
+// The type can be easily swapped without needing to change anything else in the implementation.
+#[derive(
```

Comment:
Can we make AuthorityIdentifier Copy, and mention that it is a range of values from 0 to committee.size() - 1?

Reply:
I've considered this, but thought `Copy` would be quite restrictive in case we need to change the type - but since I see no opposition on using the index the way we do I can derive as Copy

Reply:
I've updated the code to use the `Copy` for the `AuthorityIdentifier`